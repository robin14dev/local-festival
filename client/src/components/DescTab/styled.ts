import styled from "styled-components";

export const Container = styled.section`
  height: 100%;
  width: 100%;
  overflow-y: auto;
  border-radius: 0 0 1rem 1rem;
  padding: 1rem;
  padding-bottom: 5rem;

  h1 {
    font-size: 2rem;
    margin-top: 42px;
    margin-bottom: 21px;
  }
  h2 {
    color: #878585;

    font-weight: 600;
    font-size: 22px;
    line-height: 27px;
  }

  h2 + h2 {
    margin-top: 11px;
  }

  h3 {
    color: #878585;
    img {
      height: 22px;
      margin-right: 22px;
    }
  }

  @media (max-width: 485px) {
    display: none;
  }
`;
export const Overview = styled.p`
  margin-top: 43px;
  margin-bottom: 21px;
  line-height: 1.5;
  word-break: "keep-all";
  text-indent: "0.5rem";
  white-space: "pre-line";
`;
export const Contacts = styled.div`
  svg {
    height: 100%;
    width: auto;
    margin-right: 0.5rem;
  }
  button {
    display: flex;
    justify-content: center;
    align-items: center;
    font-weight: 500;
    font-size: 1rem;

    color: #878585;
  }
  button + button {
    margin-left: 1rem;
  }
  div {
    height: 1.8rem;
  }
`;
export const Links = styled.div`
  display: flex;
  margin-top: 1.5rem;
`;
