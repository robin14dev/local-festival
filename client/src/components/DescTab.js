import React from 'react';
import styled from 'styled-components';
import moment from 'moment';
import { useParams } from 'react-router-dom';
import phoneImg from '../assets/phone.png';
import linkImg from '../assets/link.png';

const Wrapper = styled.div`
  height: 100%;
  width: 100%;
  overflow-y: auto;
  /* border: 1px 0 0 0 solid black; */
  /* margin: 1rem; */
  border-radius: 0 0 1rem 1rem;
  padding: 1rem;

  h1 {
    margin-top: 42px;
    margin-bottom: 21px;
  }
  h2 {
    color: #878585;

    font-weight: 600;
    font-size: 22px;
    line-height: 27px;
  }

  h2 + h2 {
    margin-top: 11px;
  }

  p {
    margin-top: 43px;
    margin-bottom: 21px;
  }

  h3 {
    color: #878585;
    img {
      height: 22px;
      margin-right: 22px;
    }
  }

  @media (max-width: 485px) {
    display: none;
  }
`;

const DescTab = ({ festival }) => {
  // 탭메뉴의 상태가 detail(상세정보)이면 특정 행사정보 렌더링
  //             review면 review 컴포넌트 렌더링
  console.log('desktab!!!!');
  let params = useParams();
  console.log('Desktop!!', params);

  console.log(festival);
  const { location, overview, tel, title, homepageUrl, startDate, endDate } =
    festival;

  // let parsedUrl = '';
  // if (homepageUrl !== 'undefined') {
  //   parsedUrl = homepageUrl.split(' ')[1].slice(5).replace(/"/g, '');
  // } else {
  //   parsedUrl = null;
  // }
  return (
    <>
      <Wrapper>
        <h1>{title}</h1>
        <h2> {location}</h2>
        <h2>
          {moment(startDate, 'YYYY.MM.DD').format('YYYY년 MM월 DD일')} ~{' '}
          {moment(endDate, 'YYYY.MM.DD').format('YYYY년 MM월 DD일')}
        </h2>

        <p style={{ wordBreak: 'keep-all', textIndent: '0.5rem' }}>
          {overview}
        </p>

        <h3>
          <img src={phoneImg} alt="phoneImg" /> 문의{' '}
          <a href="tel:{tel}">{tel}</a>
        </h3>

        <h3>
          <img src={linkImg} alt="linkImg" /> 홈페이지{' '}
          <a target="_blank" rel="noreferrer">
            링크
          </a>
        </h3>
      </Wrapper>
    </>
  );
};

export default DescTab;
