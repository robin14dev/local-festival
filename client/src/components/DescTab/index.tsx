import React from "react";
import moment from "moment";

import { ReactComponent as InstagramIcon } from "../../assets/instagram.svg";
import { ReactComponent as YoutubeIcon } from "../../assets/youtube.svg";
import { ReactComponent as HomepageIcon } from "../../assets/homepage.svg";
import { ReactComponent as PhoneIcon } from "../../assets/phone.svg";
import { Contacts, Container, Links, Overview } from "./styled";
import { Link } from "react-router-dom";

type DescTabProps = {
  festival: FestivalItem;
};
type RegexObj = { [key: string]: RegExp };

const regex: { [key: string]: RegExp } = {
  official: /https?:\/\/(?!www.instagram\.com|www.youtube\.com)[\w\-\./]+/g,
  instagram: /https?:\/\/(www.instagram\.com)[\w\-\.\/]+/g,
  youtube: /https?:\/\/(www.youtube\.com)[\w\-\.\/]+/g,
};

const DescTab = ({ festival }: DescTabProps) => {
  const { location, overview, tel, title, startDate, endDate } = festival;

  const homepageUrl =
    "https://www.google.com   https://www.instagram.com/sefes   https://www.youtube.com/efse";

  const urlCollection: { [key: string]: string } = {
    official: "",
    instagram: "",
    youtube: "",
  };
  const getMatchedUrl = (homepageUrl: string, regexObj: RegexObj) => {
    for (const url in regexObj) {
      const urlRegex = regexObj[url];
      if (urlRegex.test(homepageUrl)) {
        const matchedUrl = homepageUrl.match(urlRegex) as string[];
        urlCollection[url] = matchedUrl[0];
      }
    }
  };
  const openLink = (url: string): void => {
    window.open(url);
  };
  if (homepageUrl) {
    getMatchedUrl(homepageUrl, regex);
  }

  const {
    official: officialLink,
    instagram: instagramLink,
    youtube: youtubeLink,
  } = urlCollection;

  return (
    <Container data-testid="DescTab">
      <h1>{title}</h1>
      <h2> {location}</h2>
      <h2>
        <span>
          {moment(startDate, "YYYY.MM.DD").format("YYYY년 MM월 DD일")}
        </span>
        <span>~</span>
        <span>{moment(endDate, "YYYY.MM.DD").format("YYYY년 MM월 DD일")}</span>
      </h2>

      <Overview>
        {overview
          .replace(/<(\/br|br)([^>]*)>/gi, "\n")
          .replace(/&lt;/gi, "<")
          .replace(/&gt;/gi, ">")}
      </Overview>

      {/**
       * Link가 url를 상대적인 주소로 인식하기 때문에 https도 빼주어야되고 www.google.com으로만 넣어줘야됨
       */}

      <Contacts>
        <div>
          <a href={`tel:${tel}`}>
            <PhoneIcon />
            {tel}
          </a>
        </div>

        <Links>
          {officialLink && (
            <a href={officialLink} target="_blank" rel="noopener noreferrer">
              <HomepageIcon /> 공식 홈페이지
            </a>
          )}
          {instagramLink && (
            <a href={instagramLink} target="_blank" rel="noopener noreferrer">
              <InstagramIcon /> 인스타그램 계정
            </a>
          )}
          {youtubeLink && (
            <a href={youtubeLink} target="_blank" rel="noopener noreferrer">
              <YoutubeIcon /> 유튜브 채널
            </a>
          )}
        </Links>
      </Contacts>
    </Container>
  );
};

export default DescTab;
