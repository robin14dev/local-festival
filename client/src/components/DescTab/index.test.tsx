import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import DescTab from ".";
import { mockFestivalItem } from "../../test/mock";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";

describe("<DescTab />", () => {
  it("데스크탑 환경에서 성공적으로 렌더링 됩니다", () => {
    /**
     * 이거는 상세페이지에서 렌더링을 해야되는거라서 이건 데스크탑 내에서만 되는걸로 해야겠네
     */
    const { container } = render(
      <MemoryRouter initialEntries={["/"]}>
        <DescTab festival={mockFestivalItem} />
      </MemoryRouter>,
    );
    expect(screen.getByTestId("DescTab")).toBeInTheDocument();
    expect(screen.getByText("축제 제목")).toBeInTheDocument();
    expect(screen.getByText("위치 정보")).toBeInTheDocument();
    expect(screen.getByText("2024년 01월 01일")).toBeInTheDocument();
    expect(screen.getByText("2024년 12월 31일")).toBeInTheDocument();
    expect(screen.getByText("축제 개요")).toBeInTheDocument();
    expect(screen.getByText("010-1234-5678")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("받아온 외부링크 별로 컴포넌트들이 성공적으로 렌더링 됩니다.", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/"]}>
        <DescTab
          festival={{
            ...mockFestivalItem,
            homepageUrl: "https://www.instagram.com/seoul",
          }}
        />
      </MemoryRouter>,
    );
    expect(screen.getByText("인스타그램 계정")).toBeInTheDocument();
  });
});
