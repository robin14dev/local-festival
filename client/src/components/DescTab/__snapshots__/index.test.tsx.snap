// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DescTab /> 데스크탑 환경에서 성공적으로 렌더링 됩니다 1`] = `
<div>
  <section
    class="sc-bcXHqe jjzluj"
    data-testid="DescTab"
  >
    <h1>
      축제 제목
    </h1>
    <h2>
       
      위치 정보
    </h2>
    <h2>
      <span>
        2024년 01월 01일
      </span>
      <span>
        ~
      </span>
      <span>
        2024년 12월 31일
      </span>
    </h2>
    <p
      class="sc-gswNZR fPbIZH"
    >
      축제 개요
    </p>
    <div
      class="sc-dkrFOg rPElY"
    >
      <div>
        <a
          href="tel:010-1234-5678"
        >
          <svg>
            phone.svg
          </svg>
          010-1234-5678
        </a>
      </div>
      <div
        class="sc-hLBbgP eLcdkS"
      >
        <a
          href="https://www.google.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg>
            homepage.svg
          </svg>
           공식 홈페이지
        </a>
        <a
          href="https://www.instagram.com/sefes"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg>
            instagram.svg
          </svg>
           인스타그램 계정
        </a>
        <a
          href="https://www.youtube.com/efse"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg>
            youtube.svg
          </svg>
           유튜브 채널
        </a>
      </div>
    </div>
  </section>
</div>
`;
